### PR TITLE
[wmco] Generate userData secret

### PR DESCRIPTION
**Pull request by mansikulkarni96**
This PR introduces generation of user data for connecting to the
windows vm through ssh configuration. This data includes a script for sending
the public key created using private key and required config for sshd.

### CODE REVIEW

File: hack/run-ci-e2e-test.sh [hunk 1]
```diff
@@ -88,7 +88,7 @@ OSDK=$(get_operator_sdk)
 # the way operator-sdk testing is done using `go test []string{}
 NODE_COUNT=${NODE_COUNT:-2}
 SKIP_NODE_DELETION=${SKIP_NODE_DELETION:-"-skip-node-deletion=false"}
-KEY_PAIR_NAME=${KEY_PAIR_NAME:-"libra"}
+KEY_PAIR_NAME=${KEY_PAIR_NAME:-"openshift-dev"}
```
Comment (aravindhp) [inline hack/run-ci-e2e-test.sh:91]: Please remember to revert before the final merge

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 2]
```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
...
+	publicKey, err := ssh.NewPublicKey(&privateKey.PublicKey)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate public key: %v", err)
+	}
+	pubKeyBytes := ssh.MarshalAuthorizedKey(publicKey)
+	return string(pubKeyBytes[:]), nil
+}
+
+// retrievePrivateKey retrieves private key in format required for generating
+// public key using the privateKeyPath
```
Comment (aravindhp) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:436]: /retrievePrivateKey retries the private key from the privateKeyPath in a format required for generating the public key`

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 3]
```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
...
+func retrievePrivateKey() (*rsa.PrivateKey, error) {
+	privateKeyBytes, err := ioutil.ReadFile(wkl.PrivateKeyPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find private key from path: %v with: %v", wkl.PrivateKeyPath, err)
+	}
+
+	privateKeyBlock, _ := pem.Decode(privateKeyBytes)
```
Comment (aravindhp) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:443]: Is it safe to ignore the remaining data?

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 4]
```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
...
+	$passwordConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+	$authFileConf = (Get-Content -path C:\ProgramData\ssh\sshd_config) -replace 'AuthorizedKeysFile __PROGRAMDATA__/ssh/administrators_authorized_keys','#AuthorizedKeysFile __PROGRAMDATA__/ssh/administrators_authorized_keys'
+	$authFileConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+	$pubKeyLocationConf = (Get-Content -path C:\ProgramData\ssh\sshd_config) -replace 'Match Group administrators','#Match Group administrators'
+	$pubKeyLocationConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+	Restart-Service sshd
+	New-item -Path $env:USERPROFILE -Name .ssh -ItemType Directory -force
+	echo "` + r.publicKey + `"| Out-File $env:USERPROFILE\.ssh\authorized_keys -Encoding ascii
+	</powershell>
+	<persist>true</persist>`),
```
Comment (aravindhp) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:497]: Please align this under `"userData"...`

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 5]
```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
...
+		},
+	}
+
+	// check if the userDataSecret already exists
+	found := &corev1.Secret{}
+	err := r.client.Get(context.TODO(), kubeTypes.NamespacedName{Name: userDataSecret.Name, Namespace: userDataSecret.Namespace}, found)
+	if err != nil {
+		log.Info("Creating a new Secret", "Secret.Namespace", userDataSecret.Namespace, "Secret.Name", userDataSecret.Name)
+		err = r.client.Create(context.TODO(), userDataSecret)
```
Comment (aravindhp) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:506]: This should only happen if the error returned is a "not found" one.

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 6]
```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
...
+	Start-Service sshd
+	$pubKeyConf = (Get-Content -path C:\ProgramData\ssh\sshd_config) -replace '#PubkeyAuthentication yes','PubkeyAuthentication yes'
+	$pubKeyConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+ 	$passwordConf = (Get-Content -path C:\ProgramData\ssh\sshd_config) -replace '#PasswordAuthentication yes','PasswordAuthentication yes'
+	$passwordConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+	$authFileConf = (Get-Content -path C:\ProgramData\ssh\sshd_config) -replace 'AuthorizedKeysFile __PROGRAMDATA__/ssh/administrators_authorized_keys','#AuthorizedKeysFile __PROGRAMDATA__/ssh/administrators_authorized_keys'
+	$authFileConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+	$pubKeyLocationConf = (Get-Content -path C:\ProgramData\ssh\sshd_config) -replace 'Match Group administrators','#Match Group administrators'
+	$pubKeyLocationConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+	Restart-Service sshd
```
Comment (aravindhp) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:493]: Can't you set the configuration and just start the service here? I don't see any reason to start the service and restart it again here.

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 7]
```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
...
+	userDataSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "windows-user-data",
+			Namespace: "openshift-machine-api",
+		},
+		Data: map[string][]byte{
+			"userData": []byte(`<powershell>
+	$url = "https://raw.githubusercontent.com/ansible/ansible/devel/examples/scripts/ConfigureRemotingForAnsible.ps1"
+	$file = "$env:temp\ConfigureRemotingForAnsible.ps1"
+	(New-Object -TypeName System.Net.WebClient).DownloadFile($url,  $file)
```
Comment (aravindhp) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:474]: Why do we need this given we are not going to run the ansible script against this VM?

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 8]
```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
...
+	& $file
+	Add-WindowsCapability -Online -Name OpenSSH.Server~~~~0.0.1.0
+	New-NetFirewallRule -DisplayName "` + types.FirewallRuleName + `"
+	-Direction Inbound -Action Allow -Protocol TCP -LocalPort ` + types.ContainerLogsPort + ` -EdgeTraversalPolicy Allow
+	Install-PackageProvider -Name NuGet -MinimumVersion 2.8.5.201 -Force
+	Install-Module -Force OpenSSHUtils -Scope AllUsers
```
Comment (aravindhp) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:480]: Is the `AllUsers` scope needed?

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 9]
```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
...
+	Restart-Service sshd
+	New-item -Path $env:USERPROFILE -Name .ssh -ItemType Directory -force
+	echo "` + r.publicKey + `"| Out-File $env:USERPROFILE\.ssh\authorized_keys -Encoding ascii
+	</powershell>
+	<persist>true</persist>`),
+		},
+	}
+
+	// check if the userDataSecret already exists
+	found := &corev1.Secret{}
```
Comment (VaishnaviHire) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:502]: Why do we need to define a variable here?

File: test/e2e/validation_test.go [hunk 10]
```diff
@@ -278,3 +278,11 @@ func testNodeTaint(t *testing.T) {
 		assert.Equal(t, hasTaint, true, "expected Windows Taint to be present on the Node: %s", node.GetName())
 	}
 }
+
+// testUserData tests if the userData created in the 'openshift-machine-api' namespace is valid
+func testUserData(t *testing.T) {
+	found := &corev1.Secret{}
```
Comment (VaishnaviHire) [inline test/e2e/validation_test.go:284]: same as above

File: deploy/olm-catalog/windows-machine-config-operator/manifests/windows-machine-config-operator.clusterserviceversion.yaml [hunk 11]
```diff
@@ -85,6 +85,13 @@ spec:
           - signers
           verbs:
           - approve
+        - apiGroups:
+          - machine.openshift.io
```
Comment (VaishnaviHire) [inline deploy/olm-catalog/windows-machine-config-operator/manifests/windows-machine-config-operator.clusterserviceversion.yaml:89]: Not sure why this `apiGroup` is used

Comment (mansikulkarni96) [reply deploy/olm-catalog/windows-machine-config-operator/manifests/windows-machine-config-operator.clusterserviceversion.yaml:89]: it was generated as a part of csv generation for the role modification

Comment (VaishnaviHire) [reply deploy/olm-catalog/windows-machine-config-operator/manifests/windows-machine-config-operator.clusterserviceversion.yaml:89]: Sorry. I am not sure why the `machine.openshift.io` apiGroup is used in `role.yaml`

Comment (mansikulkarni96) [reply deploy/olm-catalog/windows-machine-config-operator/manifests/windows-machine-config-operator.clusterserviceversion.yaml]: sorry missed that, corrected it.

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go (hunk 7 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go]: kept it for winrm config, removing it as we dont need it in operator

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 12]
```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
...
+			Namespace: "openshift-machine-api",
+		},
+		Data: map[string][]byte{
+			"userData": []byte(`<powershell>
+	$url = "https://raw.githubusercontent.com/ansible/ansible/devel/examples/scripts/ConfigureRemotingForAnsible.ps1"
+	$file = "$env:temp\ConfigureRemotingForAnsible.ps1"
+	(New-Object -TypeName System.Net.WebClient).DownloadFile($url,  $file)
+	& $file
+	Add-WindowsCapability -Online -Name OpenSSH.Server~~~~0.0.1.0
+	New-NetFirewallRule -DisplayName "` + types.FirewallRuleName + `"
```
Comment (ravisantoshgudimetla) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go]: nit: I'd just define the FirewallRuleName and ContainerLogsPort here, 2 more vars which are not dependent on WNI's types.

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go (hunk 9 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go]: removing this one, but keeping the one in test as I am asserting on it.

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 13]
```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
...
+	}
+
+	privateKey, ok := parsedKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("failed to cast private key as rsa key type: %v", err)
+	}
+	return privateKey, nil
+}
+
+// createUserDataSecret creates a secret in 'openshift-machine-api' namespace
```
Comment (ravisantoshgudimetla) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:460]: nit: call out that the name of the secret, we're interested in creating - `windows-user-data`?

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 14]
```diff
@@ -398,3 +431,93 @@ func (e *reconcileError) reason() string {
...
+			Namespace: "openshift-machine-api",
+		},
+		Data: map[string][]byte{
+			"userData": []byte(`<powershell>
+			Add-WindowsCapability -Online -Name OpenSSH.Server~~~~0.0.1.0
+			$firewallRuleName = "ContainerLogsPort"
+			$containerLogsPort = "10250"
+			New-NetFirewallRule -DisplayName $firewallRuleName -Direction Inbound -Action Allow -Protocol TCP -LocalPort $containerLogsPort -EdgeTraversalPolicy Allow
+			Install-PackageProvider -Name NuGet -MinimumVersion 2.8.5.201 -Force
+			Install-Module -Force OpenSSHUtils -Scope AllUsers
```
Comment (mansikulkarni96) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:489]: note: will remove  -Scope AllUsers in next push, avoiding push in between reviews.

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 15]
```diff
@@ -398,3 +413,76 @@ func (e *reconcileError) reason() string {
...
+		return nil, errors.Wrapf(err, "unable to parse private key: %v", wkl.PrivateKeyPath)
+	}
+	return signer, nil
+}
+
+// createUserDataSecret creates a secret 'windows-user-data' in 'openshift-machine-api'
+// namespace. This secret will be used to inject cloud provider user data for creating
+// windows machines
+func (r *ReconcileWindowsMachineConfig) createUserDataSecret() error {
+	pubKeyBytes := ssh.MarshalAuthorizedKey(r.signer.PublicKey())
```
Comment (aravindhp) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:435]: Please add check for `r.signer`.

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 16]
```diff
@@ -398,3 +413,76 @@ func (e *reconcileError) reason() string {
...
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse private key: %v", wkl.PrivateKeyPath)
+	}
+	return signer, nil
+}
+
+// createUserDataSecret creates a secret 'windows-user-data' in 'openshift-machine-api'
+// namespace. This secret will be used to inject cloud provider user data for creating
+// windows machines
+func (r *ReconcileWindowsMachineConfig) createUserDataSecret() error {
```
Comment (aravindhp) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:434]: Please keep this along with rest of the `ReconcileWindowsMachineConfig` i.e. before `ReconcileError` struct and methods.

File: pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go [hunk 17]
```diff
@@ -398,3 +413,76 @@ func (e *reconcileError) reason() string {
 func newReconcileError(reason string, err error) *reconcileError {
 	return &reconcileError{degradationReason: reason, err: err}
 }
+
+// createSigner creates a signer using the private key from the privateKeyPath
+func createSigner() (ssh.Signer, error) {
```
Comment (aravindhp) [inline pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go:418]: Please move this before the `ReconcileError` struct and methods.

File: test/e2e/validation_test.go [hunk 18]
```diff
@@ -278,3 +279,35 @@ func testNodeTaint(t *testing.T) {
 		assert.Equal(t, hasTaint, true, "expected Windows Taint to be present on the Node: %s", node.GetName())
 	}
 }
+
+// createSigner creates a signer using the private key
```
Comment (aravindhp) [inline test/e2e/validation_test.go:283]: `...private key retrieved from the secret`

**Comment by ravisantoshgudimetla** [comment]
/hold cancel

since #75 merged

**Pull request merged**